utils/retry/retry.go

```diff
@@ -0,0 +1,60 @@
+package retry
+
+import (
+	"context"
+	"time"
+
+	"github.com/rs/zerolog"
+)
+
+// BackoffExponential makes maxRetry attempts to executed function f. After the nth attempt,
+// we wait retryMilliseconds*2^n ms before retrying. Returns an error after
+// maxRetry unsuccessful attempts. It will short circuit if ctx is Done.
+func BackoffExponential(ctx context.Context, retryFuncName string, f func(context.Context) error, maxRetry int, retryMilliseconds int, logger zerolog.Logger) bool {
+	wait := time.Duration(retryMilliseconds) * time.Millisecond
+	for attempt := 1; attempt <= maxRetry; attempt++ {
+		err := f(ctx)
+		if err != nil {
+			logger.Warn().Err(err).Str("retrying_func", retryFuncName).Msgf("attempt %d/%d failed", attempt, maxRetry)
+
+			select {
+			case <-time.After(wait):
+				wait <<= 1
+				continue
+			case <-ctx.Done():
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+// WithTimeout attempts to execute function f with a retryMilliseconds sleep in between attempts
+// until the function f either succeeds, context is done or timeout has been reached.
+func WithTimeout(ctx context.Context, retryFuncName string, f func(context.Context) error, timeoutDuration time.Duration, retryMilliseconds int, logger zerolog.Logger) bool {
+	ticker := time.Tick(time.Duration(retryMilliseconds) * time.Millisecond)
+	attempt := 0
+
+	ctxWithTimeOut, cancel := context.WithTimeout(ctx, timeoutDuration)
+
+	defer cancel()
+
+	var err error
+	for {
+		select {
+		case <-ticker:
```

Comment:
I would suggest using `time.After` here as well, so the waiting mechanics are consistent with `BackoffExponential`. `Ticker` and `After` behave a bit differently (I think both are OK here, but we should use the same one for both methods).